horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -982,4 +980,150 @@ public QueryResult queryDataSet(Integer datasetId, String jsonpath, boolean arra
+                              }
+                           }
+                        }
+                     }
+                     array.append("]");
+
+                     DataSet ds = null;
+                     try {
+                        if (t.function == null || t.function.isBlank()) {
+                           ds = new DataSet(run.start, run.stop
+                              , "", test.id
+                              , Util.toJsonNode(array.toString()), run
+                              , 0, run.owner, run.access);
+                        } else {
+                           try {
+                              StringBuilder code = new StringBuilder();
+                              code.append("const __obj = { __obj == "+array.toString()+" }").append(";\n");
+                              code.append("const __func = ").append(t.function).append(";\n");
+                              code.append("__func(__obj);\n");
+                              String result = execute (test, run, code.toString(), "transformation").trim();
+                              ds = new DataSet(run.start, run.stop
+                                 , "", test.id
+                                 , Util.toJsonNode(result), run
+                                 , 0, run.owner, run.access);
+                           } finally {
+                              ds = new DataSet(run.start, run.stop
+                                 , "", test.id, run.data
+                                 , run, 0, run.owner, run.access);
+                           }
+                        }
+                     } finally {
+                        try (@SuppressWarnings("unused") CloseMe h1 = roleManager.withRoles(em, role)){
+                           ds.persist();
+                           Util.publishLater(tm, eventBus, DataSet.EVENT_NEW, ds);
+                        }
+                     }
+                  }
+               }
+            }
+         }
+         if (!matched || runSchemas.isEmpty()) {
+            new DataSet(run.start, run.stop, "",
+                  test.id, Util.toJsonNode(EMPTY_ARRAY), run, 0, run.owner, run.access)
+                  .persist();
+         }
+      }
+   }
+
+   private String execute(Test test, Run run, String jsCode, String name) {
+      ByteArrayOutputStream out = new ByteArrayOutputStream();
+      try (org.graalvm.polyglot.Context context = org.graalvm.polyglot.Context.newBuilder("js").out(out).err(out).build()) {
+         context.enter();
+         try {
+            Value value = context.eval("js", jsCode);
+            return value.toString();
+         } catch (PolyglotException e) {
+            log.errorf("Failed to execute transformation for run %d and test %d", run.id, test.id);
+            return null;
+         } finally {
+            if (out.size() > 0) {
+               logTransformationMessage(test, run, TransformationLog.WARN, "transformation", "something went wrong :P");
```

Comment:
It is nice that you already started working on logging of execution. All errors (failures to execute the JSONPath query, syntax errors in the JS function...) should be logged, too.
Beware that if you encounter an error in JDBC the transaction will probably already start rolling back, and it won't be possible to simply persist another entity (that would only throw another exception) - the errors should be logged in a separate transaction. Luckily all you need to do for that is annotate the method with
```
@Transactional(Transactional.TxType.REQUIRES_NEW)
@WithRoles(extras = Roles.HORREUM_SYSTEM)
```
(not sure if this works with private methods, but it definitely works for package-private).

Reply:
ok, let's see how that progresses